Mapping wizard step1 Form

AparnaKarve opened the pull request:
Form for Infrastructure Mapping Wizard / Step 1

- Introduces `redux-form`
- Implements Mapping Wizard / Step 1 form
- Implements validation in the form for the `Required` field
- blocks stepping further into the wizard if form has validation errors
- destroys the state of the form on Cancel/Close

Next Button disabled due to validation errors in the form -

<img width="896" alt="screen shot 2018-02-15 at 4 58 33 pm" src="https://user-images.githubusercontent.com/1538216/36288816-81e609da-1271-11e8-91df-36928c5c4c3f.png">

----------------

Next Button enabled since the form is valid -

<img width="898" alt="screen shot 2018-02-15 at 4 56 57 pm" src="https://user-images.githubusercontent.com/1538216/36288825-92445e44-1271-11e8-8adf-918005ca6228.png">

----------------

Error displayed on the field since the user went past the required field -

<img width="890" alt="screen shot 2018-02-15 at 4 57 16 pm" src="https://user-images.githubusercontent.com/1538216/36288906-ec516418-1271-11e8-8849-52caabd81323.png">


Fixes #29 #45 #46 #47

AparnaKarve commented:
/cc @priley86

ohadlevy commented on app/javascript/common/formUtilities/validation.js:
```diff
@@ -0,0 +1,23 @@
+export const required = value => (value ? undefined : 'Required');
+
```
Should those be extracted?

AparnaKarve replied:
We won't need this code once we start using `redux-form-validators`

ohadlevy commented on app/javascript/common/formUtilities/validation.js:
```diff
@@ -0,0 +1,23 @@
+    : undefined;
+
+export const alphaNumeric = value =>
+  value && /[^a-zA-Z0-9 ]/i.test(value)
```
Maybe consider using the redux form validator npm package?

AparnaKarve replied:
Will look into using redux form validator npm package. Thanks.

AparnaKarve replied:
https://github.com/priley86/miq_v2v_ui_plugin/pull/42#discussion_r167996209

ohadlevy commented on app/javascript/components/GeneralInfrastructureMapping/index.js:
```diff
@@ -0,0 +1,32 @@
+import { required } from '../../common/formUtilities/validation';
+
+const GeneralInfrastructureMappingContainer = props => (
+  <form className="form-horizontal">
```
Does it make sense to have a form component? Maybe in pf-react?

AparnaKarve replied:
I believe we already have a `Form` component in pf-react - I could use that.

ohadlevy commented on app/javascript/components/GeneralInfrastructureMapping/index.js:
```diff
@@ -0,0 +1,32 @@
+      name="name"
+      label={__('Name')}
+      component={renderField}
+      validate={[required]}
```
I think we need to pass a required flag, so we can draw the asterisk next to required fields?

AparnaKarve replied:
- In Classic UI we do not mark the Required fields with asterisk, but instead use a red-border for the Required field and also specify `Required` underneath the field (as shown in the above screenshot)
- In Foreman, we mark the Required fields with asterisk, and no red borders and no text underneath

So should we combine both styles here?
An asterisk + a red border box with `Required` text underneath

AparnaKarve replied:
Tried using the npm package - `redux-form-validators` (https://github.com/priley86/miq_v2v_ui_plugin/pull/42#discussion_r167739751), and noticed that it adds it's own Help text - `is required`, as shown below.
![image](https://user-images.githubusercontent.com/1538216/36172852-062140fa-10bc-11e8-9552-691155cf2938.png)

(Probably not a good idea to use `redux-form-validators`  for `Required` fields)

priley86 commented:
this is looking pretty good to me!

* would like to check on what components need to be included in componentRegistry (i'm not positive on this), but maybe @ohadlevy will know.
* hopefully we can keep the file structure/test structure somewhat similar for redux-form connected components as we do other redux connected components (domain level file structuring). I'm not positive on this yet or not...

michaelkro commented on app/javascript/common/formUtilities/renderField.js:
```diff
@@ -0,0 +1,41 @@
+  controlId: PropTypes.string,
+  type: PropTypes.string,
+  meta: PropTypes.object
+};
```
This looks like a duplicate file?
https://github.com/priley86/miq_v2v_ui_plugin/pull/42/files#diff-bc9a1f3d9130bcc35d98fb9c3a857ca3R1

AparnaKarve replied:
Yes, probably leftover because of the shuffle.
I have deleted the duplicate. Thanks.

mturley commented on app/javascript/components/index.js:
```diff
@@ -1,4 +1,5 @@
 import MappingWizardContainer from '../react/screens/App/Overview/screens/MappingWizard';
+// import MappingWizardGeneralStepContainer from '../react/screens/App/Overview/screens/MappingWizard/components/MappingWizardGeneralStep';
```
We should clean up these commented-out lines before merging, if you weren't doing so already.

AparnaKarve replied:
Yes, I left the comments in there to basically underline the fact that `componentRegistry` (and the related code like this one) does not work with Redux form.

I will be removing them.

AparnaKarve replied:
Addressed in d3da630

priley86 replied:
i am fine w/ removing those comments for now ;)

mturley commented on app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js:
```diff
@@ -14,10 +16,23 @@ const ModalWizard = props => {
   const onFinalStep = activeStepIndex === numSteps - 1;
+
+  const formHasErrors = () =>
+    find(formContainer, form => form.syncErrors) !== undefined;
```
Do we need to use lodash `find()` when we can use `Array.prototype.find()` now with babel? I'm fine with either, just noticing this. We use lodash enough elsewhere that it doesn't matter much.

AparnaKarve replied:
Let's keep `lodash`

mturley commented on app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js:
```diff
@@ -14,10 +16,23 @@ const ModalWizard = props => {
+  const formHasErrors = () =>
+    find(formContainer, form => form.syncErrors) !== undefined;
+
+  const disableNextStep = formHasErrors();
```
This function definition with only one immediate call could be simplified to a variable assignment:

```
const formHasErrors = find(formContainer, form => form.syncErrors) !== undefined;
const disableNextStep = formHasErrors;
```

AparnaKarve replied:
Addressed in d3da630

mturley commented on app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js:
```diff
@@ -14,10 +16,23 @@ const ModalWizard = props => {
+
+  const onCancel = () => {
+    // Delete/Destroy all wizard forms here
+    delete formContainer.mappingWizardGeneralStep;
```
Calling `delete` on a property of one of your `props`? hmm... we should never be mutating our props. This should probably instead be some kind of call to a callback from higher up that changes the formContainer prop from above. @priley86 what do you think here?

mturley replied:
I think there should be a redux action for form cancellation, this should just dispatch that. A reducer should handle the mutation of formContainer, and not by mutating it directly but by constructing an copy of it with the changes.  i.e., in the reducer:

```
const newFormContainer = { ...oldFormContainer, mappingWizardGeneralStep: null };
```

Or if you want to use one of my new helpers, `filterKeys`:

```
import { filterKeys } from 'patternfly-react';
const newFormContainer = filterKeys(oldFormContainer, key => key !== 'mappingWizardGeneralStep');
```

Alternatively, it looks like you might get a `resetForm()` prop injected that you can use instead: https://redux-form.com/6.0.0-alpha.4/docs/faq/howtoclear.md/

AparnaKarve replied:
I went through this exercise with `resetForm()` but had some difficulties making it work

Will give it a try again and let you know what I find.

AparnaKarve replied:
I went with Approach D in https://redux-form.com/6.0.0-alpha.4/docs/faq/howtoclear.md/, which seems to be working great.

Addressed in f44fa53

priley86 replied:
after discussing w/ you guys just now... i think we can just inject the form inputs into the MappingWizard via the same redux-form store keys... (so inputs can be read there as well as the sub form)... then when all forms become valid....we can construct the POST body for the mapping in MappingWizard. MappingWizard becomes the primary state manager I suppose.

mturley commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/MappingWizardBody.js:
```diff
@@ -23,6 +24,11 @@ const MappingWizardBody = props => {
   const mappingWizardClustersStepContainer = componentRegistry.markup(
     'MappingWizardClustersStepContainer'
   );
+  // This may not be required for Redux form
```
More commented out code to remove before we merge

AparnaKarve replied:
Addressed in d3da630

mturley commented on app/javascript/react/screens/App/Overview/screens/formUtilities/renderField.js:
```diff
@@ -0,0 +1,41 @@
+import PropTypes from 'prop-types';
+import { Form, Grid } from 'patternfly-react';
+
+export const renderField = ({
```
As this is a React component, it should probably be capitalized, and perhaps be a "noun" name instead of a "verb" name. `FormField`?

AparnaKarve replied:
Addressed in d3da630

mturley commented:
I’m happy to help with debugging if you share the errors you’re running
into. If you continue to be stuck on that, feel free to push the broken
changes and we can take a look, we can always revert them.  I guess we
shouldn’t break the PR for testing, so maybe if you do that use a different
branch.

On Thu, Feb 15, 2018 at 4:30 PM Aparna Karve <notifications@github.com>
wrote:

> *@AparnaKarve* commented on this pull request.
> ------------------------------
>
> In
> app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js
> <https://github.com/priley86/miq_v2v_ui_plugin/pull/42#discussion_r168613843>
> :
>
> >    } = props;
>    const onFirstStep = activeStepIndex === 0;
>    const onFinalStep = activeStepIndex === numSteps - 1;
> +
> +  const formHasErrors = () =>
> +    find(formContainer, form => form.syncErrors) !== undefined;
> +
> +  const disableNextStep = formHasErrors();
> +
> +  const onCancel = () => {
> +    // Delete/Destroy all wizard forms here
> +    delete formContainer.mappingWizardGeneralStep;
>
> I went through this exercise with resetForm() but had some difficulties
> making it work
>
> Will give it a try again and let you know what I find.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/priley86/miq_v2v_ui_plugin/pull/42#discussion_r168613843>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxjuxmyNm3uJG-BKUq81M86zdmpa4C-ks5tVKHjgaJpZM4SCvor>
> .
>

AparnaKarve commented:
@mturley I have addressed your feedback. Thank you for the thorough review :)

I have also updated the form validation per patternfly guidelines.
New screenshots posted in the top comment.

priley86 commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardGeneralStep/index.js:
```diff
@@ -0,0 +1,34 @@
+import { Form } from 'patternfly-react';
+import { FormField } from '../../../formUtilities/FormField';
+
+const MappingWizardGeneralStep = props => (
```
I think the "component" itself can be split into "MappingWizardGeneralStep.js"... but fine w/ this for now... we have the concept of "connected" inside index.js and the "unconnected" outside... This makes testing easier (if you see Overview tests )

mturley replied:
I agree with this, leaving a note here so I remember to address it if I beat @AparnaKarve to it.

AparnaKarve replied:
Will look into this shortly

priley86 commented:
@mturley do you mind testing this and sharing any more thoughts?  I'll try to review it once more tomorrow AM too...

ohadlevy commented:
Sorry for the late response, for i18n for form-validators see https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/reduxFormI18n.js

mturley commented:
Going to merge this in a sec to unblock the next piece, and address the straggler feedback tracked by those issues. Just waiting for my compiler, to make sure it's all good before I merge. Code looks good.

mturley commented:
I'm hitting issues with my Ruby environment while trying to test this. Unrelated to the changes here, and the code looks good to me. If anyone else can test this feel free to merge

AparnaKarve commented on app/javascript/components/index.js:
```diff
@@ -16,6 +17,13 @@ export const coreComponents = [
+  //   type: MappingWizardGeneralStepContainer,
+  //   data: {},
+  //   store: true
+  // },
```
Looks like I missed a spot - I can take care of this.

AparnaKarve commented:
@priley86 @mturley Addressed the above feedback  in 8318307
(which takes care of #45 #46 #47)

AparnaKarve commented:
> Sorry for the late response, for i18n for form-validators see https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/reduxFormI18n.js

@ohadlevy No worries. I found in the `redux-form-validators` doc that you can also customize validation messages like this -
```html
<Field
  name="name"
  label={__('Name')}
  required
  component={FormField}
  validate={[required({ msg: __('Required') })]}
  type="text"
/>
```

Implemented in 5800aa65ddbe06f430bffe3c5bea7fe176d4ce50

priley86 commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/MappingWizardBody.js:
```diff
@@ -2,6 +2,7 @@ import React from 'react';
 import PropTypes from 'prop-types';
```
Let's convert this to the following for now:

```
import React from 'react';
import PropTypes from 'prop-types';
import ModalWizard from '../../components/ModalWizard';
import componentRegistry from '../../../../../../components/componentRegistry';
import MappingWizardGeneralStep from '../MappingWizard/components/MappingWizardGeneralStep';

// TODO remove these, they are space fillers
const t = str => (
  <div align="center">
    <h1>TODO: {str}!</h1>
  </div>
);
const todo = str => (
  <div>
    {t(str)}
    {t(str)}
    {t(str)}
    {t(str)}
    {t(str)}
  </div>
);

class MappingWizardBody extends React.Component {
  shouldComponentUpdate(nextProps, nextState) {
    return (
      JSON.stringify(this.props) !== JSON.stringify(nextProps) ||
      JSON.stringify(this.state) !== JSON.stringify(nextState)
    );
  }
  render() {
    const mappingWizardClustersStepContainer = componentRegistry.markup(
      'MappingWizardClustersStepContainer'
    );
    return (
      <ModalWizard.Body
        {...this.props}
        loadingTitle={__('Loading Infrastructure Mappings...')}
        loadingMessage={__('This may take a minute.')}
        steps={[
          {
            title: __('General'),
            render: () => <MappingWizardGeneralStep />,
            onClick: () => console.log('on step 1 click')
          },
          {
            title: __('Clusters'),
            render: () => mappingWizardClustersStepContainer,
            onClick: () => console.log('on step 2 click')
          },
          {
            title: __('Datastores'),
            render: () => todo('Datastore Mappings Form'),
            onClick: () => console.log('on step 3 click')
          },
          {
            title: __('Networks'),
            render: () => todo('Network Mappings Form'),
            onClick: () => console.log('on step 4 click')
          },
          {
            title: __('Results'),
            render: () => todo('Display Progress and Results'),
            onClick: () => console.log('on step 5 click')
          }
        ]}
      />
    );
  }
}

MappingWizardBody.propTypes = {
  loaded: PropTypes.bool,
  activeStepIndex: PropTypes.number,
  activeStep: PropTypes.string
};

MappingWizardBody.defaultProps = {
  loaded: false,
  activeStepIndex: 0,
  activeStep: '1'
};

export default MappingWizardBody;

```

AparnaKarve replied:
Good find! Fixed in 8e95bcc

priley86 commented:
@AparnaKarve can do add the above for now to prevent the infinite loop? Looks like redux-form is not so happy...
https://github.com/erikras/redux-form/issues/2629#issuecomment-297268290

I think we are not so happy w/ it either... haha.

priley86 commented:
merge this for now to aid in other integrations. @AparnaKarve thanks very much for your diligence!
